File: CHUONG_5/5.15_ViewPager_Exercise/app/src/main/java/com/example/viewpager/ui/fragment/ZodiacFragment.java
package com.example.viewpager.ui.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.viewpager.data.datasource.ZodiacLocalDataSourceImpl;
import com.example.viewpager.data.model.Zodiac;
import com.example.viewpager.databinding.FragmentZodiacBinding;
import com.example.viewpager.ui.adapter.ZodiacPagerAdapter;
import com.google.android.material.tabs.TabLayoutMediator;

public class ZodiacFragment extends Fragment {
    private FragmentZodiacBinding binding;
    private ZodiacPagerAdapter adapter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = FragmentZodiacBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupViewPager();
    }

    private void setupViewPager() {
        ZodiacLocalDataSourceImpl dataSource = new ZodiacLocalDataSourceImpl();
        adapter = new ZodiacPagerAdapter(this);
        adapter.submitList(dataSource.getZodiacs());
        binding.viewPager.setAdapter(adapter);
        new TabLayoutMediator(binding.tabLayout, binding.viewPager,
                (tab, position) -> {
                    Zodiac zodiac = adapter.getZodiacAtPosition(position);
                    if (zodiac != null) {
                        tab.setText(zodiac.getName());
                    }
                }).attach();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: CHUONG_5/5.15_Zodiac_kotlin/app/src/main/java/com/example/zodiac/ui/fragment/ZodiacFragment.kt
package com.example.zodiac.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.example.zodiac.R
import androidx.fragment.app.viewModels
import androidx.lifecycle.ViewModelProvider
import com.example.zodiac.databinding.FragmentZodiacBinding
import com.example.zodiac.ui.ZodiacViewModel
import com.example.zodiac.ui.adapter.ZodiacPagerAdapter
import com.google.android.material.tabs.TabLayoutMediator

/**
 * A simple [Fragment] subclass.
 * Use the [ZodiacFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class ZodiacFragment : Fragment() {
    private var _binding: FragmentZodiacBinding? = null
    private val binding get() = _binding!!

    private lateinit var adapter: ZodiacPagerAdapter
    private var mediator: TabLayoutMediator? = null

    private val viewModel: ZodiacViewModel = ViewModelProvider(this)[ZodiacViewModel::class.java]

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentZodiacBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        adapter = ZodiacPagerAdapter(this)
        binding.apply {
            viewPager.adapter = adapter
            TabLayoutMediator(tabLayout, viewPager) { tab, position ->
                tab.text = adapter.getTitle(position)
            }.attach()
        }

        viewModel.zodiacs.observe(viewLifecycleOwner) {
            adapter.submitList(it)
        }
    }
}